zookeeper-server/src/test/java/org/apache/zookeeper/server/quorum/QuorumDigestTest.java

```diff
@@ -0,0 +1,191 @@
+        Assert.assertEquals(0L, getMismatchDigestCount());
+    }
+
+    @Test
+    public void testDigestMatchesWithAsyncRequests() throws Exception {
+
+        int leader = servers.findLeader();
+
+        final ZooKeeper client = servers.zk[leader];
+        final AtomicBoolean stopped = new AtomicBoolean(true);
+        final String prefix = "/testDigestMatchesWithAsyncRequests";
+
+        // start a thread to send requests asynchronously,
+        Thread createTrafficThread = new Thread () {
+            @Override
+            public void run() {
+                int i = 0;
+                while (!stopped.get()) {
+                    String path = prefix + "-" + i;
+                    client.create(path, path.getBytes(), ZooDefs.Ids.OPEN_ACL_UNSAFE,
+                            CreateMode.PERSISTENT, new StringCallback() {
+                        @Override
+                        public void processResult(int rc, String path,
+                                Object ctx, String name) {
+                            // ignore the result
+                        }
+                    }, null);
+                    try {
+                        Thread.sleep(10);
+                    } catch (InterruptedException e) { /* ignore */ }
+                }
+            }
+        };
+        createTrafficThread.start();
+
+        // shutdown a follower and observer
+        List<Integer> targets = Arrays.asList(
+                servers.findAnyFollower(), servers.findAnyObserver());
+        stopServers(targets);
+
+        // start the follower and observer to have a diff sync
+        startServers(targets);
+
+        // make sure there is no digest mismatch
+        Assert.assertEquals(0L, getMismatchDigestCount());
+
+        // stop the leader
+        targets = Arrays.asList(leader);
+        stopServers(targets);
+        startServers(targets);
+
+        // make sure there is no digest mismatch
+        Assert.assertEquals(0L, getMismatchDigestCount());
+
+        stopped.set(true);
+    }
+
+    private void stopServers(List<Integer> sids) throws InterruptedException {
+        for (int sid : sids) {
+            if (sid != -1) {
+                servers.mt[sid].shutdown();
+                waitForOne(servers.zk[sid], States.CONNECTING);
+            }
+        }
+    }
+
+    private void startServers(List<Integer> sids) throws InterruptedException {
+        for (int sid : sids) {
+            servers.mt[sid].start();
+            waitForOne(servers.zk[sid], States.CONNECTED);
+        }
+    }
+
+    private void triggerOps(int sid, String prefix) throws Exception {
+        TxnLogDigestTest.performOperations(servers.zk[sid], prefix);
+        servers.restartClient(sid, null);
+        waitForOne(servers.zk[sid], States.CONNECTED);
+    }
+
+    private void triggerSync(boolean snapSync) throws Exception {
+        if (snapSync) {
+            System.setProperty(LearnerHandler.FORCE_SNAP_SYNC, "true");
```

Comment:
In my opinion we should reset this systen property at the end and maybe even at the begin of each test.
Otherwise we are not sure about the current value before and after the test

Reply:
@eolivelli we've already doing clearProperty in the tearDown.